pkg/api/apiloader.go

```diff
@@ -42,8 +42,15 @@ func (a *Apiloader) DeserializeContainerService(contents []byte, validate bool,
 	version := m.APIVersion
 	service, err := a.LoadContainerService(contents, version, validate, existingContainerService)
 	if service == nil || err != nil {
-		log.Infof("Error returned by LoadContainerService: %+v. Attempting to load container service using LoadContainerServiceForAgentPoolOnlyCluster", err)
-		service, err = a.LoadContainerServiceForAgentPoolOnlyCluster(contents, version, validate)
+		service, matchedAgentPoolOnly, agentPoolErr := a.LoadContainerServiceForAgentPoolOnlyCluster(contents, version, validate)
```

Comment:
My recommendation would be to update isAgentPoolOnlyClusterFromProperties to check if MasterProfile is present. If it is not present call: LoadContainerServiceForAgentPoolOnlyCluster else call: LoadContainerService.

Regarding: 
 > This occurs because if an apimodel fails validation then it is retried as an agent-pool-only apimodel, and if it fails validation against that schema, then it is always the agent-pool-only validation error that is returned. 

This is not entirely correct from ACS Engine perspective since generate is a command, error is printed and not swallowed. But I understand that it could be confusing. ACS Engine is commonly cmds are commonly used only for ContainerService type (and not agent pool only type) so it makes sense to optimize for that.